### PR TITLE
Incremental

### PR DESCRIPTION
**Pull request by MusinaPolina**

### CODE REVIEW

File: src/commonMain/kotlin/org/kosat/Solver.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+     * In general, the Solver implementations construct the model on each call to [getModel].
+     * The model could have the large size, so make sure to call this method only once.
+     *
+     * **Note:** the solver should be in the SAT state.
+     * Solver return the latest model (cached)
+     * even when the solver is already not in the SAT state (due to possibly new added clauses),
+     * but it is advisable to query the model right after the call to [solve] which returned `true`.
+     */
+    fun getModel(): List<Lit>
+}
```
Comment (AlekseyVlasov) [inline src/commonMain/kotlin/org/kosat/Solver.kt:70]: new line EOF

File: src/commonMain/kotlin/org/kosat/CDCL.kt [hunk 2]
```diff
@@ -68,10 +75,36 @@ class CDCL(private var clauses: MutableList<MutableList<Int>>, private val varsN
     // list of unit clauses to propagate
     private val units: MutableList<Int> = mutableListOf()
 
+    private var assumptions: List<Int> = emptyList()
+
+    private fun clearTrail() {
+        while (trail.isNotEmpty()) {
+            delVariable(trail.removeLast())
+        }
+    }
```
Comment (AlekseyVlasov) [inline src/commonMain/kotlin/org/kosat/CDCL.kt:84]: Не хотим ли мы оставлять нулевой level в trail?

File: src/commonMain/kotlin/org/kosat/CDCL.kt [hunk 3]
```diff
@@ -5,11 +5,18 @@ import kotlin.math.abs
...
+    return if (solver.solve()) solver.getModel() else null
 }
 
+class CDCL(val clauses: MutableList<MutableList<Int>>) {
+    var varsNumber = 0
+        private set
+
+    init {
+        varsNumber = clauses.flatten().let { all -> if (all.isNotEmpty()) all.maxOf { abs(it) } else 0 }
+    }
```
Comment (AlekseyVlasov) [inline src/commonMain/kotlin/org/kosat/CDCL.kt:18]: Don't we want to get number of var from user?

File: src/commonMain/kotlin/org/kosat/CDCL.kt [hunk 4]
```diff
@@ -68,10 +75,30 @@ class CDCL(private var clauses: MutableList<MutableList<Int>>, private val varsN
     // list of unit clauses to propagate
     private val units: MutableList<Int> = mutableListOf()
 
+    private var assumptions: List<Int> = emptyList()
+
+    private fun clearTrail(until: Int) {
+        while (trail.isNotEmpty() && vars[trail.last()].level > until) {
```
Comment (AlekseyVlasov) [inline src/commonMain/kotlin/org/kosat/CDCL.kt:81]: 0 by default?

File: src/commonMain/kotlin/org/kosat/CDCL.kt [hunk 5]
```diff
@@ -68,10 +75,30 @@ class CDCL(private var clauses: MutableList<MutableList<Int>>, private val varsN
...
+        }
+        clearTrail(0)
+        return  result
+    }
+
     fun solve(): List<Int>? {
-        removeUselessClauses()
 
-        // extreme cases
+        // extremal cases
```
Comment (AlekseyVlasov) [inline src/commonMain/kotlin/org/kosat/CDCL.kt:101]: extreme cases

File: src/commonMain/kotlin/org/kosat/CDCL.kt [hunk 6]
```diff
@@ -68,10 +75,30 @@ class CDCL(private var clauses: MutableList<MutableList<Int>>, private val varsN
...
+                clearTrail(0)
+                return null
+            }
+        }
+        clearTrail(0)
+        return  result
+    }
+
     fun solve(): List<Int>? {
-        removeUselessClauses()
```
Comment (AlekseyVlasov) [inline src/commonMain/kotlin/org/kosat/CDCL.kt:72]: why not?

File: src/commonMain/kotlin/org/kosat/CDCL.kt [hunk 7]
```diff
@@ -131,19 +181,29 @@ class CDCL(private var clauses: MutableList<MutableList<Int>>, private val varsN
...
-            var cnt = 0
-            val clauseVars = clause.map { litIndex(it) }
-            for (ind in trail.lastIndex downTo 0) { // want to watch on last 2 literals from trail for conflict clause
-                if (trail[ind] in clauseVars) {
-                    cnt++
-                    watchers[trail[ind]].add(index)
-                    if (cnt == 2) {
-                        return
-                    }
+            addForLastInTrail(2, clause, index)
```
Comment (AlekseyVlasov) [inline src/commonMain/kotlin/org/kosat/CDCL.kt:194]: what if clause already unsat?

File: src/commonMain/kotlin/org/kosat/CDCL.kt [hunk 8]
```diff
@@ -1,15 +1,23 @@
...
+    return if (solver.solve()) solver.getModel() else null
 }
 
+class CDCL(val clauses: MutableList<MutableList<Int>>, initNumber: Int = 0) {
+    var varsNumber = initNumber
+        private set
+
+    init {
+        varsNumber = max(initNumber, clauses.flatten().let { all -> if (all.isNotEmpty()) all.maxOf { abs(it) } else 0})
+    }
```
Comment (AlekseyVlasov) [inline src/commonMain/kotlin/org/kosat/CDCL.kt:19]: comment for hard logic

File: src/commonMain/kotlin/org/kosat/CDCL.kt [hunk 9]
```diff
@@ -68,27 +76,57 @@ class CDCL(private var clauses: MutableList<MutableList<Int>>, private val varsN
...
+    fun solveWithAssumptions(currentAssumptions: List<Int> = emptyList()): List<Int>? {
+        assumptions = currentAssumptions
+        val result = solve()
+        assumptions.forEach {
+            if (getStatus(it) == VarStatus.FALSE) {
+                return null
+            }
+        }
+        return  result
+    }
```
Comment (AlekseyVlasov) [inline src/commonMain/kotlin/org/kosat/CDCL.kt:96]: Shouldn't we clear assumptions before return? In case we call solve right after solveWithAssumptions

**Review by AlekseyVlasov — COMMENTED** [review]
Minor changes left

File: src/commonMain/kotlin/org/kosat/CDCL.kt (hunk 9 above)
Comment (MusinaPolina) [reply src/commonMain/kotlin/org/kosat/CDCL.kt:96]: Really

**Review by AlekseyVlasov — APPROVED** [review]

**Pull request merged**